### PR TITLE
Enhance naming of enums

### PR DESCRIPTION
**Pull request by jodastephen**
Add `NamedEnum` interface, to mark standard enums
Add `EnumNames` to ensure more lenient and consistent parsing
Enhanced performance of parse and format

### CODE REVIEW

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/AccrualOnDefaultFormula.java [hunk 1]
```diff
@@ -44,31 +47,18 @@ private AccrualOnDefaultFormula(String name) {
...
    * @return the type
    * @throws IllegalArgumentException if the name is not known
    */
   @FromString
-  public static AccrualOnDefaultFormula of(String uniqueName) {
-    ArgChecker.notNull(uniqueName, "uniqueName");
-    String str = CaseFormat.UPPER_CAMEL.to(CaseFormat.UPPER_UNDERSCORE, uniqueName);
-    if (str.endsWith("I_S_D_A")) {
-      str = "ORIGINAL_ISDA";
-    }
```
Comment (brianweller89) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/AccrualOnDefaultFormula.java]: Assume we are happy for this endsWith logic to be completely removed?

File: modules/product/src/main/java/com/opengamma/strata/product/bond/FixedCouponBondYieldConvention.java [hunk 2]
```diff
@@ -55,22 +58,23 @@ private FixedCouponBondYieldConvention(String name) {
...
+   * @param name  the name to parse
    * @return the type
    * @throws IllegalArgumentException if the name is not known
    */
   @FromString
-  public static FixedCouponBondYieldConvention of(String uniqueName) {
-    ArgChecker.notNull(uniqueName, "uniqueName");
-    return valueOf(uniqueName.replace('-', '_').replace("/", "").toUpperCase(Locale.ENGLISH));
+  public static FixedCouponBondYieldConvention of(String name) {
+    return NAMES.parse(name);
```
Comment (brianweller89) [inline modules/product/src/main/java/com/opengamma/strata/product/bond/FixedCouponBondYieldConvention.java]: Again we're losing some custom parsing logic here.

File: modules/collect/src/main/java/com/opengamma/strata/collect/named/EnumNames.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  /**
+   * Parses the standard external name for an enum.
+   * 
+   * @param name  the external name
+   * @return the enum value
+   */
+  public T parse(String name) {
+    ArgChecker.notNull(name, "name");
```
Comment (jodastephen) [inline modules/collect/src/main/java/com/opengamma/strata/collect/named/EnumNames.java]: True, but this way users get a better error message

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/AccrualOnDefaultFormula.java (hunk 1 above)
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/AccrualOnDefaultFormula.java]: That was a hack that is now supported properly.

File: modules/product/src/main/java/com/opengamma/strata/product/bond/FixedCouponBondYieldConvention.java (hunk 2 above)
Comment (jodastephen) [inline modules/product/src/main/java/com/opengamma/strata/product/bond/FixedCouponBondYieldConvention.java]: That was a hack that is now supported properly.

**Pull request merged**